packages/@glimmer/application-test-helpers/src/app-builder.ts

```diff
@@ -242,6 +236,20 @@ export class AppBuilder<T extends TestApplication> {
   }
 }
 
+const CAPABILITIES = {
```

Comment:
Rationale?

Reply:
TypeScript would have inferred the “wrong” type here: it would think capabilities is a pojo with whatever fields you have here. If you made a typo or missed a field, you will only get the error much later when you try to use this value by passing it to another API, and often a confusing error. If you had said this is `CAPABILITIES: ComponentCapabilities` here typescript can point out any errors or omission inline instead of cascading it down. It also makes future upgrades easier

Reply:
I think in practice it would get caught by the `getComponentCapabilities()` hook immediately below, but I have no objections.

Is the rule that we _always_ require annotating types for object literal data structures that cannot be inferred? I don't know if we're consistent about this in Glimmer.js/Glimmer VM. If that's not the rule of thumb, what is?

Reply:
@tomdale yes, I think that's a good practice.

When you're defining the capabilities list here, you really do want TypeScript to give you an error if you didn't supply a correct capabilities list, right here.

The problem is that object literals never "cannot be inferred". When you write an object literal, it always has the type of whatever you typed, and therefore will always type check.

Another way to think about it is that type inference is useful if you know that a particular type will type check, and you want to avoid typing the type. It's unhelpful to rely on type inference in a situation where the types might be wrong in the first place.

Reply:
https://www.typescriptlang.org/play/?ssl=1&ssc=1&pln=8&pc=8#code/MYewdgzgLgBAZiEAuGAxRMC8MDe9EoBEARgIYBOhANDGeUWQF6EwC+A3AFCcCWYUAU3JxSwAWgx4EyGNHJ8A5m25wArmGBQe4eAAoA+inQgAlLk6sVu6SfZA

In this playground, if you delete the type for `const foo`, you will stop getting the type error for the extra field. This is what I meant by making upgrades easier: if we removed a capability we want to get an error here.

Reply:
@chancancode You wouldn't get an error when passing it as an argument, but you would get an error caused by the unused property when you return from a method whose return type is specified: https://www.typescriptlang.org/play/#code/MYewdgzgLgBAZiEMC8MDe9EC4YCIBGAhgE64A0MRxOBhAXrjAL4DcAUGwJZhQCmxcQsF4wAYonSYQOaMW4BzZhzgBXMMCidwMALaEA1r3EgAFAEocx9Gxi2YxXlBXEwkhNLxVylEjSINmdiZlNQ0tVzgTAH1LRDNrYLZI9zMWIA

But in any case, I agree this gives you better checking, I have no objections, and have updated the code to specify the type explicitly.